Src/BufferHelpers.cpp

```diff
@@ -0,0 +1,281 @@
+
+    return S_OK;
+}
+
+
+//--------------------------------------------------------------------------------------
+_Use_decl_annotations_
+HRESULT DirectX::CreateTextureFromMemory(
+    ID3D12Device* device,
+    ResourceUploadBatch& resourceUpload,
+    size_t width,
+    DXGI_FORMAT format,
+    const D3D12_SUBRESOURCE_DATA& initData,
+    ID3D12Resource** texture,
+    D3D12_RESOURCE_STATES afterState)
+{
+    if (!texture)
+        return E_INVALIDARG;
+
+    *texture = nullptr;
+
+    if (!device || !width || !initData.pData)
+        return E_INVALIDARG;
+
+    if (width > D3D12_REQ_TEXTURE1D_U_DIMENSION || width > UINT32_MAX)
+    {
+        DebugTrace("ERROR: Resource dimensions too large for DirectX 12 (1D: size %zu)\n", width);
+        return HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED);
+    }
+
+    auto desc = CD3DX12_RESOURCE_DESC::Tex1D(format, static_cast<UINT64>(width), 1u, 1u);
+
+    CD3DX12_HEAP_PROPERTIES heapProperties(D3D12_HEAP_TYPE_DEFAULT);
+
+    ComPtr<ID3D12Resource> res;
+    HRESULT hr = device->CreateCommittedResource(
+        &heapProperties,
+        D3D12_HEAP_FLAG_NONE,
+        &desc,
+        D3D12_RESOURCE_STATE_COPY_DEST,
+        nullptr,
+        IID_GRAPHICS_PPV_ARGS(res.GetAddressOf()));
+    if (FAILED(hr))
+        return hr;
+
+    try
+    {
+        resourceUpload.Upload(res.Get(), 0, &initData, 1);
+
+        resourceUpload.Transition(res.Get(), D3D12_RESOURCE_STATE_COPY_DEST, afterState);
+    }
+    catch (com_exception e)
+    {
+        return e.get_result();
+    }
+    catch (...)
+    {
+        return E_FAIL;
+    }
+
+    *texture = res.Detach();
+
+    return S_OK;
+}
+
+_Use_decl_annotations_
+HRESULT DirectX::CreateTextureFromMemory(
+    ID3D12Device* device,
+    ResourceUploadBatch& resourceUpload,
+    size_t width,
+    size_t height,
+    DXGI_FORMAT format,
+    const D3D12_SUBRESOURCE_DATA& initData,
+    ID3D12Resource** texture,
+    bool generateMips,
+    D3D12_RESOURCE_STATES afterState)
+{
+    if (!texture)
+        return E_INVALIDARG;
+
+    *texture = nullptr;
+
+    if (!device || !width || !height
+        || !initData.pData || !initData.RowPitch)
+        return E_INVALIDARG;
+
+    if ((width > D3D12_REQ_TEXTURE2D_U_OR_V_DIMENSION) || (width > UINT32_MAX)
+        || (height > D3D12_REQ_TEXTURE2D_U_OR_V_DIMENSION) || (height > UINT32_MAX))
+    {
+        DebugTrace("ERROR: Resource dimensions too large for DirectX 12 (2D: size %zu by %zu)\n", width, height);
+        return HRESULT_FROM_WIN32(ERROR_NOT_SUPPORTED);
+    }
+
+    uint16_t mipCount = 1;
+    if (generateMips)
+    {
+        generateMips = resourceUpload.IsSupportedForGenerateMips(format);
+        if (generateMips)
+        {
+            mipCount = static_cast<uint16_t>(LoaderHelpers::CountMips(static_cast<uint32_t>(width), static_cast<uint32_t>(height)));
+        }
+    }
+
+    auto desc = CD3DX12_RESOURCE_DESC::Tex2D(format, static_cast<UINT64>(width), static_cast<UINT>(height), 1u, mipCount);
```

Comment:
If Tex2D accepts a UINT64, why is there a check above ensuring it's < UINT32_MAX?